### PR TITLE
The MODsuit Digital Revolution: Replaces AIs in MODsuits with pAIs

### PR DESCRIPTION
**Pull request by GoldenAlpharex**
## About The Pull Request
Replaces AIs being inserted into MODsuits with pAIs.

This was originally in /tg/'s #63743, but John ended up giving up after maintainers didn't seem to agree with the direction he wanted to go. Obviously, most of the credits go to him, and I've gotten his authorization to re-use his code upstream.

The maintainers weren't happy, because they'd rather have some much less pretty code for MMIs to be in MODsuits instead, with pAIs as a side-dish if someone wants to also support that, which I just fundamentally disagree with, so instead of never seeing this thing becoming a reality (because I doubt that code is going to be written anytime soon), it's going to be Skyrat-exclusive until they decide to do something about it.

Thus, I present a touched-up version of his PR, where pAIs do not have the ability to move the MODsuit at all unless `can_pai_move_suit` is set to `TRUE` on the mod control unit, which currently can only be done by admins in-game. I'm thinking of potentially making a program which would take the majority (or entirety) of a pAI's RAM to allow them to move the suit, but regardless, they won't be able to move the suit if the wearer is not conscious (so anything between soft crit and death, passing by unconscious), to make it so they can't be used as a mean to basically give you a second chance at life, or to be your personal paramedic.

I also made removing the pAI from the MODsuit something handled via TGUI, as can be seen here:
image

Of course,  it requires the cover to be opened to allow the user to eject the pAI and doesn't appear for the pAIs, because they shouldn't be able to get themselves out of the MODsuit against their owner's wish.

## Why It's Good For The Game
You can find most of the reasoning in PR #63743 on /tg/, it stays true here as well, but basically, long story short: pAIs don't have inherent AA, access to all comms channels and binary, and don't affect the round by taking a job that's meant to interact with the station and putting it in a spot where it's basically the equivalent of being one-human'd. As a plus, it's now also a lot more intuitive to get your little buddy in your MODsuit, because you can actually pick and choose personalities that would want to be in a MODsuit, which you couldn't really do with positronic brains that you then turned into AIs.

Eventually I'd want to possibly even make it so you can select from a list of reasons why you're requesting a pAI, so people know what they're signing up for when someone's requesting a pAI, but that's a story for another day.

Side-note: I did ask John for his consent before making this PR. If Fulp wants to port this, they're absolutely free to, I mean, John was the one that made most of this, I simply brought everything up to speed and integrated it in TGUI, and made a few fixes. Not sure they'll do it, though, as I heard they reset their repo every so often, so I'm not sure how this'd work out for them.

## Changelog

:cl: GoldenAlpharex for the touch-ups, the QoL and the TGUI, JohnFulpWillard for the initial code
balance: AIs controllers MODsuits cannot move an incapacitated or dead owner anymore.
refactor: Modsuits have been refactored to allow pAIs to enter them, replacing AIs.
qol: Taking your pAI out of your MODsuit is now done via the MODsuit's interface menu rather than with an empty hand.
qol: pAIs will now get a bit more feedback on the state of the MODsuit when it gets activated or deactivated.
/:cl:

### CODE REVIEW

**Comment by Zenitheevee** [comment]
Better than using AI's? Maybe, Definitely less resource intensive than making an AI shell
More balanced? _eh_, Balance is probably not exactly the thing i'd put on the table here? AI's shunted into a modsuit can still do AI stuff so i'd say atleast a solid _maybe_ heading towards yes.
More inline for Bay Rigsuit type styling? Definitely.

**Comment by GoldenAlpharex** [comment]
> Better than using AI's? Maybe, Definitely less resource intensive than making an AI shell More balanced? _eh_, Balance is probably not exactly the thing i'd put on the table here? AI's shunted into a modsuit can still do AI stuff so i'd say atleast a solid _maybe_ heading towards yes. More inline for Bay Rigsuit type styling? Definitely.

Having an AI in your MODsuit gave you literally access to all comms, so it was either "Give a non-head literally Captain-level of comms access" or restrict them to heads of staff, which absolutely sucks.
Plus, it also served as a literal all-access, so I'd say this is a *lot* more balanced than having AIs in MODsuits.

**Comment by Zenitheevee** [comment]
> > Better than using AI's? Maybe, Definitely less resource intensive than making an AI shell More balanced? _eh_, Balance is probably not exactly the thing i'd put on the table here? AI's shunted into a modsuit can still do AI stuff so i'd say atleast a solid _maybe_ heading towards yes. More inline for Bay Rigsuit type styling? Definitely.
> 
> Having an AI in your MODsuit gave you literally access to all comms, so it was either "Give a non-head literally Captain-level of comms access" or restrict them to heads of staff, which absolutely sucks. Plus, it also served as a literal all-access, so I'd say this is a _lot_ more balanced than having AIs in MODsuits.

Fair point on that. That asside it also seemed like a whole ass process to get an AI for a modsuit, Constructing an AI shell and *Explicitely* stating they're gonna be shunted into a modsuit... Not something someone who *genuinely* wants to play AI would want.

**Comment by CandleJaxx** [comment]
as long as this doesn't actually hard-remove the ai's ability to be in a modsuit, this should be fine? there's no reason to not allow it when there's already so much built into it

**Comment by GoldenAlpharex** [comment]
> as long as this doesn't actually hard-remove the ai's ability to be in a modsuit, this should be fine? there's no reason to not allow it when there's already so much built into it

It does, because truth be told, the AIs were only put in there because Fikou couldn't make pAIs work without a big refactor (which conveniently happened after MODsuits were merged).
The reasons behind it are all listed in John's PR, and summarized in mine (I honestly didn't feel like writing a novel), but I'm just overall incredibly not satisfied with AIs in MODsuits, and having to choose between the busted state they're in right now, or pAIs, which are much more credible and balanced, I'm going with pAIs. The side-effect of this is that you'll have a much easier time choosing who goes into your MODsuit, since you're the one picking from the list, which I'd say is very, very nice. You also get notified if someone submits their personality, so you can just hang around with it in your pocket and download it whenever you hear the ding.

**Comment by ghost** [comment]
My question is, how many of the pAI's abilities remain?
Can they use their photo mode? Music player? **Doorjack?**
If so, based as hell. If not, irrelevant.

Either way, +1 from me, code looks sound (though I wish this could have been done without having to comment out code ;-;)

**Comment by CandleJaxx** [comment]
> It does, because truth be told, the AIs were only put in there because Fikou couldn't make pAIs work without a big refactor (which conveniently happened after MODsuits were merged).
> The reasons behind it are all listed in John's PR, and summarized in mine (I honestly didn't feel like writing a novel), but I'm just overall incredibly not satisfied with AIs in MODsuits, and having to choose between the busted state they're in right now, or pAIs, which are much more credible and balanced, I'm going with pAIs. The side-effect of this is that you'll have a much easier time choosing who goes into your MODsuit, since you're the one picking from the list, which I'd say is very, very nice. You also get notified if someone submits their personality, so you can just hang around with it in your pocket and download it whenever you hear the ding.

if that's the case, this seems like something you should do upstream, no? it refractors the entire set of modsuits code, and it'd be way easier to modify it down the line if you pushed it up there, if that's what fikou wanted in the first place

**Comment by GoldenAlpharex** [comment]
> if that's the case, this seems like something you should do upstream, no? it refractors the entire set of modsuits code, and it'd be way easier to modify it down the line if you pushed it up there, if that's what fikou wanted in the first place

The whole reason I did it here is because Upstream didn't want it, it's written in the body of the PR. I didn't directly link the PR because I'm not about creating drama.

**Comment by GoldenAlpharex** [comment]
> My question is, how many of the pAI's abilities remain? Can they use their photo mode? Music player? **Doorjack?** If so, based as hell. If not, irrelevant.
> 
> Either way, +1 from me, code looks sound (though I wish this could have been done without having to comment out code ;-;)

I haven't tested the modules, if I'm entirely honest.

**Comment by CandleJaxx** [comment]
> The whole reason I did it here is because Upstream didn't want it, it's written in the body of the PR. I didn't directly link the PR because I'm not about creating drama.

that's  fair, but if that's the case, you shouldn't cite "fikou wanted it" as a reason for implementing this- it really should be put to a vote in silicon policy or something like that, first.

**Comment by GoldenAlpharex** [comment]
> that's fair, but if that's the case, you shouldn't cite "fikou wanted it" as a reason for implementing this- it really should be put to a vote in silicon policy or something like that, first.

No, Fikou wanted it, there was just a maintainer that didn't and as such has blocked it. And, no, I don't think this warrants a silicon policy vote, for there's no policy about AIs in MODsuits, and from an administrative standpoint, it's a bit of a nightmare if we'd have to actually crack on people using them. This is simply much safer than using AIs, I don't have to worry about people getting AA and all comms as regular crewmembers, and as such, it's making it a *lot* more accessible for your average crewmember, which is a good thing in my opinion. It's fun with little drawbacks over having it in your pocket instead.

**Comment by GoldenAlpharex** [comment]
> My question is, how many of the pAI's abilities remain? Can they use their photo mode? Music player? **Doorjack?** If so, based as hell. If not, irrelevant.
> 
> Either way, +1 from me, code looks sound (though I wish this could have been done without having to comment out code ;-;)

Okay, so, I tested: Photo mode "works", but because they're inside of something the picture is black, wouldn't recommend. Loudness booster works juuuust fine (very nice to have badass music when you run through the halls with your MODsuit). Doorjack sadly doesn't work because it checks if the pAI's direct `loc` (so what it's inside of) is living, and a MODsuit isn't alive. I could probably fix that, but I'd do that Upstream, and it would just trickle down to here after some time.

**Comment by gonenoculer5** [comment]
Why not just have both options available. Why _remove_ the functionality of AI's in modsuits as they were originally implemented, instead of adding the ability to have either or? I recall it being a thing where AI's were head only, which is cool and sounds reasonable enough to me, but PAI's could be for any and all crew since they dont carry the funny gold card abilities. 

Edit: Especially since creation of AI's requires approval from the RD and such too, afaik.

**Comment by GoldenAlpharex** [comment]
> Why not just have both options available. Why _remove_ the functionality of AI's in modsuits as they were originally implemented, instead of adding the ability to have either or? I recall it being a thing where AI's were head only, which is cool and sounds reasonable enough to me, but PAI's could be for any and all crew since they dont carry the funny gold card abilities.
> 
> Edit: Especially since creation of AI's requires approval from the RD and such too, afaik.

I literally answered it before, AIs in MODsuits are busted balance-wise and it's not something I want to be encouraged.

**Comment by SpookyTheFox** [comment]
Well that's sad.
This is a near total loss of the feature.

File: code/modules/mod/mod_actions.dm
Comment (Gandalf2k15) [inline code/modules/mod/mod_actions.dm]: I really don't like the amount of non-modular changes to this file, it's gonna make major refactors a pain in the ass

File: code/modules/mod/mod_actions.dm
Comment (Gandalf2k15) [inline code/modules/mod/mod_actions.dm]: Cant you just make a new action and leave this  in, but replace the instance where it gives the action?

File: code/modules/mod/mod_activation.dm
Comment (Gandalf2k15) [inline code/modules/mod/mod_activation.dm:170]: Could probably be a trinary operator single line change

File: modular_skyrat/modules/modsuit_pai/code/mod_pai.dm
Comment (Gandalf2k15) [inline modular_skyrat/modules/modsuit_pai/code/mod_pai.dm:24]: Can maybe make these visible alerts

File: modular_skyrat/modules/modsuit_pai/code/mod_pai.dm
Comment (Gandalf2k15) [inline modular_skyrat/modules/modsuit_pai/code/mod_pai.dm:56]: What if the actions in the suit change? The action button will remain?

**Review by Gandalf2k15 — CHANGES_REQUESTED** [review]
Overall seems fine, though, this amount of non-modular changes will be an issue later.

Also, can't it support both AI's and PAI's?

File: .vscode/settings.json
Comment (GoldenAlpharex) [inline .vscode/settings.json]: Why the fuck did I commit that?
```suggestion
	}
```

File: .vscode/settings.json
Comment (GoldenAlpharex) [inline .vscode/settings.json]:
```suggestion
	}
```

File: code/modules/mod/mod_actions.dm
Comment (GoldenAlpharex) [reply code/modules/mod/mod_actions.dm]: It was just so we wouldn't have a trailing variable that didn't do anything, but it's whatever I suppose. I'll be maintaining this nonetheless, though.

File: code/modules/mod/mod_activation.dm
Comment (GoldenAlpharex) [reply code/modules/mod/mod_activation.dm:170]: It could, but only if I was leaving support for AIs, which I won't.

File: modular_skyrat/modules/modsuit_pai/code/mod_pai.dm
Comment (GoldenAlpharex) [reply modular_skyrat/modules/modsuit_pai/code/mod_pai.dm:56]: The actions won't change unless it's pinned actions, which require the pAI user to intentionally do that, and I'd assume that's already handled by the action itself, and if it's not, we can test that and I can push a fix upstream, because they'll likely have the same issue with MODsuit AIs.

File: modular_skyrat/modules/modsuit_pai/code/mod_pai.dm
Comment (GoldenAlpharex) [reply modular_skyrat/modules/modsuit_pai/code/mod_pai.dm:24]: Hmmm, I could I suppose, yeah. I hadn't really touched that part too much, though, I guess.
I'd try it out on live with just balloon alerts first, and then we'll make it a visible message if it feels necessary.

**Comment by CandleJaxx** [comment]
eh, there really should be a broader discussion over changing a core mechanic abt modsuits rather than just putting it on the git when upstream didn't want it- if it's a skyrat-specific change, those playerbases should be directly informed

File: modular_skyrat/modules/modsuit_pai/code/mod_pai.dm
Comment (GoldenAlpharex) [reply modular_skyrat/modules/modsuit_pai/code/mod_pai.dm:56]: That was fixed upstream.

**Comment by Gandalf2k15** [comment]
I am still really not set  on the non-modularity of this, it's gonna be a massive pain later on

**Comment by ghost** [comment]
Honestly, for the sake of modularity, how much of the AI code could stay and simply be inaccessible? I.E. removing the ability to place the AI into the suit and leaving the pAI added variables?

Also a few things with pAIs not liking modsuits:
Host Scan only works if the user is holding it, not if its in a storage, doubly an issue for this (also its plain bad anyways - literally makes the pAI a glorified health scanner);
Door Jack drops the cable (somewhere???) instead of at the user's feet;
Might want to make sure the Internal GPS, Atmos Sensor, Printer, and Internal Light work still too considering the above;

**Comment by Gandalf2k15** [comment]
COnflicts begin

**Pull request merged**
